Allow script to run without init system (in containers) (#38)
devexp-db/postgresql-setup: README.md

# postgresql-setup

## Requires

### BuildRequires
- m4
- docbook-utils
- help2man
- elinks (pretty README.rpm-dist)
- postgresql-server (detect current version)

### Requires
- coreutils

### Suggested BuildRequires
- util-linux (mountpoint utility)

### Maintainer's BuildRequires
- autoconf
- automake
- autoconf-archive

> ## Usage

Reviewer: I would add a note about the feature you are implementing with this PR

Author: Good idea! Will also include that.

This script is used as a wrapper around PostgreSQL initialization and upgrade
commands. It also parses init system service files and/or enviroment files to
correctly find datadir based on current system.

### Initialization
To initialize new PostgreSQL data directory use `./postgresql-setup --initdb`.

### Upgrade
To upgrade existing PostgreSQL data directory to use newer PostgreSQL version
use `./postgresql-setup --upgrade`.

If your distribution doesn't include this

> script with PostgreSQL and you are using this on your own, please update

Reviewer: Do I understand it correctly that you mean users who want to use it on other distributions?

Author: Yes, from what I understood this script is completely distro/os agnostic (as long as u have bash) and you could configure the paths.

And with the addition of not being dependent on systemd you have even more possibilities with this script. I mean, theoretically it could even be used on macOS with bash and postgresql installed (via brew, for example). Only thing that you would need is to fix the paths via autotools config (maybe something is hardcoded somewhere, but mostly just autotools).

# postgresql-setup

## Requires

### BuildRequires
- m4
- docbook-utils
- help2man
- elinks (pretty README.rpm-dist)

> - postgresql-server (detect current version)

Reviewer 2: I believe this should also list `make`

```suggestion
### BuildRequires
- make
- m4
- docbook-utils
- help2man
- elinks (pretty README.rpm-dist)
- postgresql-server (detect current version)
```

### Requires
- coreutils

### Suggested BuildRequires
- util-linux (mountpoint utility)

### Maintainer's BuildRequires
- autoconf
- automake
- autoconf-archive

## Usage
This script is used as a wrapper around PostgreSQL initialization and upgrade

> commands. It also parses init system service files and/or enviroment files to

Reviewer 2: Typo

```suggestion
commands. It also parses init system service files and/or environment files to
```

correctly find datadir based on current system.

### Initialization
To initialize new PostgreSQL data directory use `./postgresql-setup --initdb`.

### Upgrade
To upgrade existing PostgreSQL data directory to use newer PostgreSQL version
use `./postgresql-setup --upgrade`.

If your distribution doesn't include this
script with PostgreSQL and you are using this on your own, please update

> `etc/postgresql-setup/upgrade/postgresql.conf` to reflect your setup.

Reviewer 2: I'd be explicit about current folder, as this may invoke missing `/` at the beginning. 

```suggestion
`./etc/postgresql-setup/upgrade/postgresql.conf` to reflect your setup.
```

### Running without systemd/init system
Your setup might not include systemd as the init system or include any
init system at all. This would be the case in most of the base container images
for example. This script will try to compensate by parsing systemd
service file in preconfigured path directly. By default the path is
`/lib/systemd/system`.

If there is no systemd service file, or for whatever reason the script is unable
to find valid PostgreSQL data directory, you can still provide PostgreSQL data
directory path manually by using `--datadir` argument. For example when

> initializing new data directory use `./postgresql-setup --initdb --datadir=/my/path`.

Reviewer 2: `postgresql-setup` is getting exported, there's no `postgresql-setup` in `.`

```suggestion
initializing new data directory use `postgresql-setup --initdb --datadir=/my/path`.
```

This feature is most beneficial when using this script inside container images,
as it gives you the most control with least dependencies.

## Maintainer notes

> Be careful about paths. Your might need to tweak paths either in configure

Reviewer 2: Typo
```suggestion
Be careful about paths. You might need to tweak paths either in the configure
```

  files, or in code based on your environment.

> - Line 49 of `/bin/postgresql-setup` in function `builddir_source ()` has to

Reviewer 2: I don't know why, but this function doesn't even appear in my `/bin/postgresql-setup`, it gets all replaced by an empty string. Here's my `/bin/postgresql-setup`:

```
...
# Home directory of postgres user
POSTGRES_HOMEDIR=/var/lib/pgsql

# The where PostgreSQL server listens by default
PGPORT_DEF=5432

. "/usr/share/postgresql-setup/library.sh"

:

# We upgrade by default from system's default PostgreSQL installation
option_upgradefrom="postgresql"

srvsuff=
test 0 -eq 0 && srvsuff=".service"
...
```

As for the suggestion, you wanna fix that before deploying?

```suggestion
- Line 49 of `./bin/postgresql-setup.in` in function `builddir_source ()` has to
```

Author: Yes, this was indeed targeting the `.in` file. As to why it got replaced, I'll look into that as I specifically had to make this change to make it work from the working directory, instead of installing it. Maybe I got confused before, tho.
